packages/affiliates/libs/contracts.js

```diff
@@ -70,14 +70,29 @@ function GetInputLength(abi) {
   };
 }
 
-const DecodeAttribution = abi => {
+// Given a transaction, decode the attribution tag from the function. By default only suppports create.
+const DecodeAttribution = (abi, name = "create") => {
   // convert bits to hex (div by 4) and add 2 for 0x
-  const inputLength = GetInputLength(abi)("create") / 4 + 2;
-  return createTransaction => {
-    assert(createTransaction.name == "create", "Only decodes create transactions");
+  const inputLength = GetInputLength(abi)(name) / 4 + 2;
```

Comment:
Somewhat unrelated, but took a brief look at your GetInputLength implementation. One idea that may be worth toying with is seeing if you can get web3 or ethers to do some of the heavy lifting for you. web3 has a simple interface for interacting with abis: https://web3js.readthedocs.io/en/v1.3.0/web3-eth-abi.html. Ethers has a more complex, but more complete, library for it: https://docs.ethers.io/v5/single-page/#/v5/api/utils/abi/.

One idea for a more generalized implementation to decode the attribution is to use the ethers `parseTransaction` [method](https://docs.ethers.io/v5/single-page/#/v5/api/utils/abi/interface/-%23-Interface--parsing). You could basically do two operations (may require a little testing):
1. Parse the transaction in full. Get the result (or catch the error).
2. Parse the transaction with the last 20 bytes removed. Get the result (or catch the error).

My guess is that you could use the errors and/or outputs to definitively determine whether the transaction was tagged without explicitly needing to calculate the data lengths yourself.

Not sure how necessary any of this is in the short term, just thought about it as I was reading through, and I wanted to mention the abi libs if you hadn't seen them.

Reply:
@mrice32 Thanks for showing me this. One thing to note is that `parseTransaction` is currently used to parse the transaction data before tags are looked at. But I think what you are suggesting is that we slice off an address length chunk of data from the raw unparsed transaction string and see if it parses, if so, we have a tag. If it errors then parse the full intact string, and therefore no tag.

This approach works because it assumes your tags are address sized, which I wanted to avoid this assumption in case we ended up with some other kind of scheme for encoding tags (even variable length), but I think it could work with the current assumption of tag size and be quite a bit simpler to implement

The other thing is that you have pointed out an interesting observation, this works because data is encoded at the end of every transaction so you can assume the tag will always be at the end of the raw tx string. I did not realize that but its good to know.

Reply:
Yep, that's what I'm suggesting.

It seems that we're looking at the problem in two different ways:
- I'm suggesting that we infer the details of the transaction data by assuming the structure of the tag.
- I think you're suggesting that we infer the details of the tag by assuming the structure of the transaction data.

I think those are both reasonable approaches. My only minor argument for the former is that we have many types of transactions in our system that vary substantially, but our tags are fairly simple. So my intuition is that assuming that the structure of the tags is stable and simple is safer than assuming that the structure of the transaction data is at least simple (i.e. no variable-length data or ABIV2 or variable length data complexities).

Either way, this LGTM! Just wanted to mention it :)